src/network/endpoint.ml

```diff
@@ -1,13 +1,60 @@
 open Deku_consensus
+open Deku_concepts
+
+type 'a post = Post of 'a
+type get = Get
+
+type _ endpoint =
+  | Consensus : External_message.packed post endpoint
+  | Get_block_by_level : Level.t -> get endpoint
 
-type _ endpoint = Consensus : External_message.packed endpoint
 type 'a t = 'a endpoint
 type ex = Ex : _ endpoint -> ex
 
 let consensus = Consensus
 
-let of_string path =
-  match path with "/consensus" -> Some (Ex Consensus) | _ -> None
+module Parser = struct
+  let of_yojson t_of_yojson yojson =
+    try t_of_yojson yojson |> Option.some with _ -> None
+
+  let consensus path meth =
+    match (path, meth) with
+    | "/consensus", `POST -> Some (Ok (Ex Consensus))
```

Comment:
I will try to use the  https://github.com/anuragsoni/routes suggested by Ulrik
I don't know if the sumptypes will solve the issue,
I will need a function to parse a string to a sumtype and then the sumtype to the gadt 
I will refacto the parsing of the url, I am sure the suggested lib will help a lot, it can extract parameters from uri, etc...